Support distinct keyword in Postgres JSONBAgg

orf opened the pull request:
This PR adds support for a distinct argument to JSONBAgg. As there are a number of aggregations that support this I made a base class and switched the relevant aggregations to using it.

Other aggregations like `BitAnd`, `BitOr` etc support `DISTINCT` as well, but it seems somewhat redundant to add them (I think?).

charettes commented on django/contrib/postgres/aggregates/base.py:
```diff
@@ -0,0 +1,8 @@
+from django.db.models import Aggregate
+
+
+class DistinctAggregate(Aggregate):
```
I wonder if this is something we could define in `django.db.models.Aggregate` and have `Count` use it as well? What do you think @orf?

orf replied:
Yes, this is a good idea, it's not really postgres specific (other than postgres supporting it in most agg functions). I'll do that now :+1:

orf commented:
@charettes I've moved it to `django.db.models.aggregates`. It would be nice to conditionally support `DISTINCT` in standard aggregates like Min/Max etc but only for Postgres. If you have any idea on how to do this (or if it's a good idea at all) I could add support for that, but otherwise I hope it's good to merge.

charettes commented:
> It would be nice to conditionally support DISTINCT in standard aggregates like Min/Max etc but only for Postgres.

Would that perform any better? I'm trying to figure out why you'd want to use `MIN/MAX(DISTINCT)` as it would return the exact same output?

orf commented:
Yeah, on MIN/MAX it would be redundant (bad examples to choose, my bad!), but on `Avg`, `StdDev`, `Sum` and `Variance` it would change the result.

orf commented:
@charettes tests are passing, can this be merged before the end of the sprints today?

charettes commented:
@orf I'm afraid I can't commit to that.

timgraham commented:
Is there a ticket for adding `distinct` to `JSONBAgg`? It's missing the `..attribute::` documentation added in b5393028bfc939adf14d0fa5e4088cddd3b9dfa1. I think the refactoring could be done without a ticket. Could you send a separate PR for that and then rebase this branch so it's a bit cleaner?

orf commented:
There is no actual @timgraham, I just pointed out on the django bugtracker that the number of things supporting `DISTINCT` keeps growing, so perhaps a base class would be good. This grew out of that as I realized `JSONBAgg` was the last Postgres specific agg missing that functionality.

I can do it as a separate MR tomorrow at the sprints 👍

charettes commented:
@orf about `DISTINCT` `Avg` and friends this can be done in another PR. I assume we'd be merging `DistinctAggregate` into `Aggregate` and have `Aggregate.as_sql` call `connection.support_distinct_agg(self)` (to be created) if `self.distinct` and raise an exception if it doesn't support it.

orf commented:
That seems like a good way forward @charettes, but a bit out of my league right now I think. I've split the MR into #8324 @timgraham, and this one now just deals with the JSONBAgg. I've also fixed up the documentation.

orf commented:
Closed, re comment in #8324
